consensus/src/creation/mod.rs

```diff
@@ -133,86 +108,97 @@ async fn keep_processing_units_until<H: Hasher>(
             return;
         }
     };
 
-    if let Err(err) = run_creator(conf, io, starting_round).await {
+    if let Err(err) = run_creator(conf, io, keychain, starting_round).await {
         match err {
             CreatorError::OutChannelClosed(e) => {
-                warn!(target: "AlephBFT-creator", "Notification send error: {}. Exiting.", e)
+                warn!(target: LOG_TARGET, "Notification send error: {}. Exiting.", e)
             }
             CreatorError::ParentsChannelClosed => {
-                debug!(target: "AlephBFT-creator", "Incoming parent channel closed, exiting.")
+                debug!(target: LOG_TARGET, "Incoming parent channel closed, exiting.")
             }
         }
     }
 }
 
-async fn run_creator<H: Hasher>(
+async fn run_creator<H: Hasher, D: Data, MK: MultiKeychain, DP: DataProvider<D>>(
     conf: Config,
-    io: &mut IO<H>,
+    io: &mut IO<H, D, MK, DP>,
+    keychain: MK,
     starting_round: Round,
 ) -> anyhow::Result<(), CreatorError> {
-    let Config {
-        node_id,
-        n_members,
-        create_lag,
-        max_round,
-    } = conf;
+    let node_id = conf.node_ix();
+    let n_members = conf.n_members();
+    let create_delay = conf.delay_config().unit_creation_delay.clone();
+    let max_round = conf.max_round();
+    let session_id = conf.session_id();
     let mut creator = Creator::new(node_id, n_members);
+    let packer = Packer::new(keychain, session_id);
     let incoming_parents = &mut io.incoming_parents;
     let outgoing_units = &io.outgoing_units;
+    let data_provider = &mut io.data_provider;
 
-    debug!(target: "AlephBFT-creator", "Creator starting from round {}", starting_round);
+    debug!(target: LOG_TARGET, "Creator starting from round {}", starting_round);
     for round in starting_round..max_round {
         // Skip waiting if someone created a unit of a higher round.
         // In such a case at least 2/3 nodes created units from this round so we aren't skipping a
         // delay we should observe.
         let skip_delay = creator.current_round() > round;
         if !skip_delay {
-            let lag = Delay::new(create_lag(round.into()));
+            let delay = Delay::new(create_delay(round.into()));
 
-            keep_processing_units_until(&mut creator, incoming_parents, lag).await?;
+            keep_processing_units_until(&mut creator, incoming_parents, delay).await?;
         }
 
-        let (unit, parent_hashes) = create_unit(round, &mut creator, incoming_parents).await?;
-
-        trace!(target: "AlephBFT-creator", "Created a new unit {:?} at round {:?}.", unit, round);
+        let (preunit, parent_hashes) = create_unit(round, &mut creator, incoming_parents).await?;
+        trace!(target: LOG_TARGET, "Created a new preunit {:?} at round {:?}.", preunit, round);
+        let data = data_provider.get_data().await;
+        trace!(target: LOG_TARGET, "Received data.");
```

Comment:
Since this is `trace`, maybe add some details about `received data`, like `was it None or Some`.

Reply:
Good point, it turns out `Data` is apparently `Debug`, so I just print the whole thing out.